chalk-ir/src/fold.rs

```diff
@@ -62,205 +62,150 @@ pub use self::subst::Subst;
 /// ```rust,ignore
 /// let x = x.fold_with(&mut folder, 0);
 /// ```
-pub trait Folder<TF: TypeFamily, TTF: TypeFamily>:
-    FreeVarFolder<TTF> + InferenceFolder<TTF> + PlaceholderFolder<TTF> + TypeFolder<TF, TTF>
-{
-}
-
-pub trait TypeFolder<TF: TypeFamily, TTF: TypeFamily = TF> {
-    fn fold_ty(&mut self, ty: &Ty<TF>, binders: usize) -> Fallible<Ty<TTF>>;
-    fn fold_lifetime(&mut self, lifetime: &Lifetime<TF>, binders: usize)
-        -> Fallible<Lifetime<TTF>>;
-}
-
-impl<T, TF, TTF> Folder<TF, TTF> for T
-where
-    T: FreeVarFolder<TTF> + InferenceFolder<TTF> + PlaceholderFolder<TTF> + TypeFolder<TF, TTF>,
-    TF: TypeFamily,
-    TTF: TypeFamily,
-{
-}
-
-/// A convenience trait that indicates that this folder doesn't take
-/// any action on types in particular, but just recursively folds
-/// their contents (note that free variables that are encountered in
-/// that process may still be substituted). The vast majority of
-/// folders implement this trait.
-pub trait DefaultTypeFolder {}
-
-impl<T, TF, TTF> TypeFolder<TF, TTF> for T
-where
-    T: FreeVarFolder<TTF> + InferenceFolder<TTF> + PlaceholderFolder<TTF> + DefaultTypeFolder,
-    TF: TypeFamily,
-    TTF: TargetTypeFamily<TF>,
-{
+pub trait Folder<TF: TypeFamily, TTF: TargetTypeFamily<TF> = TF> {
+    /// Creates a `dyn` value from this folder. Unfortunately, this
+    /// must be added manually to each impl of Folder; it permits the
+    /// default implements below to create a `&mut dyn Folder` from
+    /// `Self` without knowing what `Self` is (by invoking this
+    /// method). Effectively, this limits impls of `Folder` to types
+    /// for which we are able to create a dyn value (i.e., not `[T]`
+    /// types).
+    fn as_dyn(&mut self) -> &mut dyn Folder<TF, TTF>;
```

Comment:
Ah, yeah, forgot about that. I tried for a while to work around it but all my workarounds were worse. We could of course just use `impl Trait` instead of `dyn Trait`, I suppose, but I"m trying to avoid that in an effort to improve compilation times.